### DISCUSSION TITLE
Ambient mode SNAT setup for health check probes

### DISCUSSION THREAD

**Discussion by mmerrill3** [istio/istio#53125] [General]
While looking at the logic to setup health check SNAT in ambient mode, where CNI creates an IPSet for 169.254.7.127/32 (v4 and v6), and adds rules, it looks like the way the rule is added at https://github.com/istio/istio/blob/d65eb93abe73c8b79156e425a09c8e9b0fd5f633/cni/pkg/iptables/iptables.go#L288, it's not really constricted to the port for the health check.  It looks like any traffic coming from a host socket (meaning any container running on the host network), can bypass the mesh, not just kubelet's health check probes.  Is this intentional?

**Comment by mmerrill3** [comment]
Source traffic that could bypass the mesh in this case, and send traffic to any port to the target pod's IP:
- daemonsets running with host network 
- processes that are running on the host itself
- pods on the same hosts that are running with host network

I think one way to fix this is to use a hash:ip,port IPSet instead of hash:IP, but CNI needs to know the ports for all the pods on the same host, which is another larger problem.  Is this just an existing larger issue with the ambient approach, in that CNI really doesn't know what ports to use for all the health probes for all the pods on the node?

**Reply by howardjohn** [reply]
This is working as intended.

Note that allowing hostNetwork traffic is not really unprecedented, ~all CNIs are doing this for network policy out of the box (https://kubernetes.io/docs/concepts/services-networking/network-policies/#networkpolicy-and-hostnetwork-pods etc).

Adding a port specification is not really meaningful for most cases. Its quite common to have health check probes on the same port the application serves other traffic on, so there is not really a meaningful difference between distinguishing based on the port or not

**Reply by mmerrill3** [reply]
Thanks for the response.   I started down this rabbit hole b/c of the comment in the function I pasted.  I'll make a documentation issue to update the comment at https://github.com/istio/istio/blob/d65eb93abe73c8b79156e425a09c8e9b0fd5f633/cni/pkg/iptables/iptables.go#L284, since it can lead user's to believe that the port of the health check is part of the match in the iptable rule.

**Reply by mmerrill3** [reply]
@howardjohn, thinking about this more, if we want to be emphatic, and only allow health checks from kubelet, we could do this with the owner module from iptables if kubelet is using a known uid or gid.  I see that as possible, and a nice additional feature to increase the effectiveness of authorization policies.
I do think that applications running on the same port as a health check is likely.  However, anecdotally, its not been my experience.
